### PR TITLE
Hardening travis 'Run tests' workflow when git-workflow caching fails

### PR DESCRIPTION
**Pull request by return42**
## What does this PR do?

Hardening travis 'Run tests' workflow. This PR is a alternative to #2517.  I recommend to read along the git comments.  First the makefile.python is polished and the requirement files are organzied.  The last commit [ensures that `make test` works as expected](https://github.com/searx/searx/pull/2517#discussion_r567240235) even if

- the python interpreter is missing
- the virtualenv exists but pyyaml is missing

## Why is this change important?

github action workflow caches the ./local directory.  Sometimes this cache can be corrupted (it is not clear how the cache gets corrupted).

- https://github.com/searx/searx/pull/2514#issuecomment-769731867
- #2517

## How to test this PR locally?

1. initial run `make travis.test`
2. manipulate the virtualenv, e.g. delete the files from the yaml module

now you have _a coroupted cache_ ....

3. run `make travis.test` again

The tests should run without fail.

### CODE REVIEW

File: requirements-dev.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
+-r requirements.txt
```
Comment (dalf) [inline requirements-dev.txt]: why ? 
It makes the setup more complicated.

Usually the way to go is:
* `pip install -r requirements.txt`
* `pip install -r requirements-dev.txt`
* `pip install -e .[test]`

File: utils/makefile.python [hunk 2]
```diff
@@ -19,6 +19,7 @@ PY       ?=3
 PYTHON   ?= python$(PY)
 PIP      ?= pip$(PY)
 PIP_INST ?= --user
+PY_PRJ_REQ ?= requirements.txt
```
Comment (dalf) [inline utils/makefile.python]: when to use `make PY_PRJ_REQ=requirements-dev.txt pyenv` and not `make pyenvinstall` ?
(related to `-r requirements.txt` in `requirements-dev.txt` )

File: Makefile [hunk 3]
```diff
@@ -251,6 +251,11 @@ test.clean:
 # travis
 # ------
 
+PHONY += travis-pyenv
+travis.test:
```
Comment (dalf) [inline Makefile]: `travis.codecov` is misleading: we don't use travis anymore, but [github actions](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions).

A generic name can be `ci.test` (so why don't have to rename the target if we move away from github actions).

Anyway it is simple effective way to make sure the CI will work !

File: utils/makefile.python [hunk 4]
```diff
@@ -97,13 +115,7 @@ quiet_cmd_pyinstall = INSTALL   $2
...
 # $2 path to folder with setup.py, this uses pip from pyenv (not OS!)
 quiet_cmd_pyenvinstall = PYENV     install $2
-      cmd_pyenvinstall = \
-	if ! cat $(PY_ENV)/requirements.sha256 2>/dev/null | sha256sum --check --status 2>/dev/null; then \
-		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install -e $2$(PY_SETUP_EXTRAS) ;\
-		sha256sum requirements*.txt > $(PY_ENV)/requirements.sha256 ;\
-	else \
-		echo "PYENV     $2 already installed"; \
-	fi
+      cmd_pyenvinstall = $(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install -e $2$(PY_SETUP_EXTRAS)
```
Comment (dalf) [inline utils/makefile.python]: Why remove the check about `requirements-dev.txt` ?

Without it, `make test` check each time the dependencies of `requirements-dev.txt`: everything since there is `-r requirements.txt`.

File: requirements-dev.txt (hunk 1 above)
Comment (return42) [reply requirements-dev.txt]: Is it? then we should implement it. ATM in branch master the workflow of `make pyenvisntall` is slightly different:

1. `pyenv` --> `pip install  -U pip wheel setuptools; pip install -r requirements.txt`
2. `pyenvinstall` --> `pip  install -e .\[test\] ; sha256sum requirements*.txt > ./local/py3/requirements.sha256`

ATM in branch master the `requirements-dev.txt` is never used by a pip command.  But `pip  install -e .\[test\]` will read from this file

But I will try to implement what you suggested.

File: utils/makefile.python (hunk 2 above)
Comment (return42) [reply utils/makefile.python]: The question should be: _In current master branch, where is `requirements-dev.txt` used?_ .. see https://github.com/searx/searx/pull/2524#discussion_r567641395

File: requirements-dev.txt (hunk 1 above)
Comment (return42) [reply requirements-dev.txt]: 'pyenv' now installs `requirements*.txt` ... see commit message https://github.com/searx/searx/pull/2524/commits/b66df00

File: utils/makefile.python (hunk 2 above)
Comment (return42) [reply utils/makefile.python]: I renamed this variable to `PYENV_REQ ?= requirements*.txt` .. a glob pattern matching requirement files.

File: Makefile (hunk 3 above)
Comment (return42) [reply Makefile]: Good point .. renamed to `ci.test` (https://github.com/searx/searx/pull/2524/commits/b53b91e1c4c7034f7777426f1679f6b68d79a487).

File: utils/makefile.python (hunk 4 above)
Comment (return42) [reply utils/makefile.python]: > Why remove the check about requirements-dev.txt ?

The sha256sum of the `requirements*.txt` is done in the `cmd_virtualenv` function.

The function `cmd_pyenvinstall` does now try to `import searx,  pkg_resources` .. installation is only done if this import fails.

**Comment by return42** [comment]
@dalf since you have merged #2517 there is only one commit 88d870b left from this PR. Do you like to have a look?  / Thanks!

File: utils/makefile.python [hunk 5]
```diff
@@ -112,15 +114,19 @@ quiet_cmd_pyinstall = INSTALL   $2
...
+quiet_cmd_pyenvinstall = PYENV     [pyenvinstall] $(PY_ENV): install $2
       cmd_pyenvinstall = \
-	if ! cat $(PY_ENV)/requirements.sha256 2>/dev/null | sha256sum --check --status 2>/dev/null; then \
-		rm -f $(PY_ENV)/requirements.sha256; \
-		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install -e $2$(PY_SETUP_EXTRAS) &&\
-		sha256sum requirements*.txt > $(PY_ENV)/requirements.sha256 ;\
+	_IMP_STATEMENT="import $(foreach var,$(wildcard $(PYOBJECTS)),$(var), ) pkg_resources"; \
+	_PRINT="print('          [pyenvinstall] pkg_resources.working_set --> resources %s' % len([True for i in pkg_resources.working_set]))"; \
+	if ! $(PY_ENV_BIN)/python -c \
+		"import sys; sys.path.pop(0); $$_IMP_STATEMENT; $$_PRINT" 2>/dev/null;\
```
Comment (dalf) [inline utils/makefile.python]: I end up to understand, but why display this information ?

Comment (return42) [reply utils/makefile.python:122]: > I end up to understand, but why display this information ?

good question .. I will remove it.

Comment (return42) [reply utils/makefile.python]: > I end up to understand, but why display this information ?

Line has been removed ..

File: docs/dev/makefile.rst [hunk 6]
```diff
@@ -36,42 +36,61 @@ Python environment
...
+   PYENV     [virtualenv] installing requirements*.txt into ./local/py3
    ...
+   Successfully installed ...
 
 With target ``pyenv`` a development environment (aka virtualenv) was build up in
 ``./local/py3/``.  To make a *developer install* of searx (:origin:`setup.py`)
 into this environment, use make target ``install``:
 
-.. code:: sh
+.. code:: txt
```
Comment (kvch) [inline docs/dev/makefile.rst]: Why did you change it to `txt`? Isn't this block showing shell output?

File: docs/dev/makefile.rst [hunk 7]
```diff
@@ -36,42 +36,61 @@ Python environment
...
-the :origin:`setup.py` file (or the requirements listed in
+If you think, something goes wrong with your ``./local`` environment or you
+change the :origin:`setup.py` file (or the requirements listed in
 :origin:`requirements-dev.txt` and :origin:`requirements.txt`), you have to call
 :ref:`make clean`.
 
+``PYENV_REQ``: requirement files (glob pattern: ``requirements*.txt``)
+  For ever reasons, if you want to install only ``requirements.txt``
+
+  .. code:: txt
```
Comment (kvch) [inline docs/dev/makefile.rst]: Same: Why is it `txt`?

File: docs/dev/makefile.rst [hunk 8]
```diff
@@ -102,10 +127,13 @@ Drop all intermediate files, all builds, but keep sources untouched.  Includes
 target ``pyclean`` which drops ./local environment.  Before calling ``make
 clean`` stop all processes using :ref:`make pyenv`.
 
-.. code:: sh
+.. code:: txt
```
Comment (kvch) [inline docs/dev/makefile.rst]: `txt`?

File: docs/dev/makefile.rst (hunk 6 above)
Comment (return42) [reply docs/dev/makefile.rst]: > Isn't this block showing shell output?

It shows the output! .. The (plain text) output of any command should not be highlighted by a bash syntax.  But thanks for the hint: it should be `text` not `txt` .. I fixed, rebased and force pushed ..

File: docs/dev/makefile.rst (hunk 7 above)
Comment (return42) [reply docs/dev/makefile.rst]: see https://github.com/searx/searx/pull/2524#discussion_r579314307

File: docs/dev/makefile.rst (hunk 8 above)
Comment (return42) [reply docs/dev/makefile.rst]: see https://github.com/searx/searx/pull/2524#discussion_r579314307

**Comment by return42** [comment]
Closed in favor of #2594

**Comment by return42** [comment]
Closed in favor of #2594
